### DISCUSSION TITLE
How to properly type a component which accepts `control` from `useForm` and renders `Controller`

### DISCUSSION THREAD

**Discussion by amankkg** [react-hook-form/react-hook-form#2700] [Q&A] [answered]
**Update**: here is an issue now #3500

I have a component wrapper for a custom input which renders a `Controller` in the end. And, I want to rely on type inference depending on passed prop `control` from `useForm`.

There is a working solution in my current project, but it has type errors.

Repo with the error https://github.com/amankkg/rhf-ts-demo (note it uses `typescript@4`).  
Codesandbox version https://codesandbox.io/s/jovial-microservice-k015c?file=/src/input.tsx

Before TypeScript v4 there was a slightly different solution without generic controller props (`Parameters<typeof Controller>[0]` instead of `ControllerProps<T, C>`) which had no type errors, but it was kinda hacky.

**Comment by bluebill1049** [comment]
codesandbox plz.

**Reply by amankkg** [reply]
Added. It looks like there is no 100% support of TypeScript that's why I created a repo.

P.S: in such cases, I prefer gitpod.io over CodeSandbox

**Comment by amankkg** [comment, marked as answer]
Linked an issue addressing this problem, also PR is proposed, so go to #3500
